### PR TITLE
[App Configuration] Support etag for kvset with new API version

### LINKED ISSUE #24121: [App Configuration - Support etag for kvset] API Review
New API Review meeting has been requested.

**Service Name**: Azure App Configuration - Azure App Configuration
**Review Created By**: Avani Gupta
**Review Date**: 06/07/2023 09:00 AM PT
**Onboarding Record**: 
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/24119
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=2698d2fd-96f9-ed11-8849-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by avanigupta**
# Data Plane API - Pull Request
AppConfiguration service wants to introduce a feature that allows users to monitor a collection of key-values for changes using ETags.

We are modifying the existing list key-values API to include “If-None-Match” and “If-Match” headers in request. Additionally, the response will contain a new “ETag” header, and the response body will also contain this ETag.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/24118

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: 
* Previous API Spec Doc: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2022-11-01-preview/appconfiguration.json
* Updated paths: /kv and /revisions

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/24121

### CODE REVIEW

**Review by JeffreyRichter — APPROVED** [review]

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json [hunk 1]
```diff
@@ -0,0 +1,2079 @@
...
+        "x-ms-long-running-operation": true
+      },
+      "patch": {
+        "tags": [
+          "Snapshots"
+        ],
+        "summary": "Updates the state of a key-value snapshot.",
+        "operationId": "UpdateSnapshot",
+        "consumes": [
+          "application/json"
```
Comment (mikekistler) [inline specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:955]: Patch should consume `application/merge-patch+json`.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json [hunk 2]
```diff
@@ -0,0 +1,2079 @@
...
+        "status": {
+          "description": "The current status of the snapshot.",
+          "type": "string",
+          "enum": [
+            "provisioning",
+            "ready",
+            "archived",
+            "failed"
+          ],
+          "x-nullable": false,
```
Comment (mikekistler) [inline specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1655]: You should avoid the use of `x-nullable: false`.  It is not needed (it is the default) and just adds noise to the API definition.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json [hunk 3]
```diff
@@ -0,0 +1,2079 @@
...
+          }
+        },
+        "composition_type": {
+          "description": "The composition type describes how the key-values within the snapshot are composed. The 'key' composition type ensures there are no two key-values containing the same key. The 'key_label' composition type ensures there are no two key-values containing the same key and label.",
+          "type": "string",
+          "enum": [
+            "key",
+            "key_label"
+          ],
+          "x-nullable": false,
```
Comment (mikekistler) [inline specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1678]: Another instance of `x-nullable: false` that should be removed. I see there are also more so please do this throughout.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json [hunk 4]
```diff
@@ -0,0 +1,2079 @@
...
+          "description": "The time that the snapshot was created.",
+          "format": "date-time",
+          "type": "string",
+          "readOnly": true
+        },
+        "expires": {
+          "description": "The time that the snapshot will expire.",
+          "format": "date-time",
+          "type": "string",
+          "x-nullable": true,
```
Comment (mikekistler) [inline specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1694]: You should avoid the use of `x-nullable: true`. Properties with no value should simply be omitted from the payload.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json [hunk 5]
```diff
@@ -0,0 +1,2079 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "2023-05-01-preview",
+    "title": "Azure App Configuration"
+  },
+  "schemes": [
+    "https"
+  ],
```
Comment (mikekistler) [inline specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:9]: You should specify a security scheme and security requirement for your operations. See the documentation
in the [Azure API Style Guide](https://github.com/Azure/azure-api-style-guide/blob/main/openapi-style-guide.md#security).

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json [hunk 6]
```diff
@@ -0,0 +1,2079 @@
...
+        "x-ms-enum": {
+          "name": "SnapshotFields",
+          "modelAsString": true
+        }
+      },
+      "x-ms-parameter-location": "method",
+      "collectionFormat": "csv"
+    },
+    "Status": {
+      "name": "Status",
```
Comment (mikekistler) [inline specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:2049]: Query parameter names should be lower camel case.
```suggestion
      "name": "status",
```

**Review by mikekistler — CHANGES_REQUESTED** [review]
Looks good.

The Spectral linter reported a few minor issues that should be easy fixes.

**Review by markweitzel — APPROVED** [review]
Per API Stewardship Board review on 7-Jun-23 this is approved for preview.
See: #24121

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 5 above)
Comment (avanigupta) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:9]: I haven't modified this part of the spec in this PR. The only changes are under /kv and /revisions paths.

**Comment by avanigupta** [comment]
@mikekistler I noticed that your comments are unrelated to the API changes in this PR. They should be part of the previous preview version that is currently being released. What would be the right way to address those?
cc @jimmyca15 since the comments are related to Snapshots.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 5 above)
Comment (jimmyca15) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:9]: @avanigupta 

Although it's not a change, if it's a requirement that was omitted in previous iterations we should try to resolve it in this one.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 6 above)
Comment (jimmyca15) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:2049]: @avanigupta 

This looks like a change we should take in this version.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 4 above)
Comment (jimmyca15) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1694]: @mikekistler 

I thought `x-nullable` was needed to denote that it was possible for the property to be omitted. Meaning, without `x-nullable`, there should always be a value for expires. 

You're saying we don't need `x-nullable` and we can still omit it from the response?

**Comment by jimmyca15** [comment]
@avanigupta we can address issues in this API version.

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 1 above)
Comment (avanigupta) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:955]: Updated, thanks!

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 2 above)
Comment (avanigupta) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1655]: Removed all occurrences of x-nullable: false

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 5 above)
Comment (avanigupta) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:9]: Added securityDefinitions.

**Review by tjprescott — APPROVED** [review]

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 4 above)
Comment (mikekistler) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1693]: That's correct. You can omit any value that is not "required". `x-nullable: true" means that [the property can be present but with the value "null"](https://github.com/Azure/autorest/tree/main/docs/extensions#x-nullable).

**Review by mikekistler — APPROVED** [review]
Thanks for the fixes! 👍

File: specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json (hunk 4 above)
Comment (avanigupta) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1693]: To resolve this issue, should I just remove `"x-nullable": true` here?

Comment (mikekistler) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json:1693]: Yes, just remove `"x-nullable": true`.

Comment (avanigupta) [reply specification/appconfiguration/data-plane/Microsoft.AppConfiguration/preview/2023-05-01-preview/appconfiguration.json]: Updated, thanks

**Pull request merged**